Store computes fees

rosalindstengle opened the pull request:
Tossing this up to make sure I'm moving in more or less the right direction. Tests mostly don't exist or don't pass - gonna keep working on those.

mrice32 commented on v1/contracts/Store.sol:
```diff
@@ -0,0 +1,60 @@
+        emit SetFixedOracleFeePerSecond(newOracleFee);
+    }
+
+    function computeRegularFee(uint startTime, uint endTime, uint pfc, bytes32 identifier) external view returns (uint regularFee, uint latePenalty) {
```
We have a new FixedPoint library: https://github.com/UMAprotocol/protocol/blob/3c77112fe7635a4519775ec4f52a73eccc4d581c/v1/contracts/FixedPoint.sol#L17. So I would take `pfc` as a FixedPoint.Unsigned and return `regularFee` and `latePenalty` as FixedPoint.Unsigned numbers.

To use this, you'll probably need to add `using FixedPoint for FixedPoint.Unsigned;` and `using FixedPoint for uint;` at the top of this contract. You'll also need to import the `FixedPoint.sol` file.

rosalindstengle commented on v1/contracts/Store.sol:
```diff
@@ -0,0 +1,60 @@
+    function computeRegularFee(uint startTime, uint endTime, uint pfc, bytes32 identifier) external view returns (uint regularFee, uint latePenalty) {
+        uint timeDiff = endTime.sub(startTime);
+
+        regularFee = pfc.mul(regularFee.sub(timeDiff));
```
Right!

mrice32 commented on v1/contracts/Store.sol:
```diff
@@ -0,0 +1,68 @@
+        emit SetFixedOracleFeePerSecond(newOracleFee);
+    }
+
+    function computeRegularFee(uint startTime, uint endTime, FixedPoint.Unsigned calldata pfc,
```
nit: breaking this function up this way doesn't adhere to solidity style - see https://solidity.readthedocs.io/en/v0.5.8/style-guide.html#function-declaration for a guide of how to do this.

TLDR: one way that you can match their suggestions is this:
```
    function computeRegularFee(uint startTime, uint endTime, FixedPoint.Unsigned calldata pfc, bytes32 identifier)
        external
        view 
        returns (FixedPoint.Unsigned memory regularFee, FixedPoint.Unsigned memory latePenalty)
    {
```

mrice32 commented on v1/contracts/StoreInterface.sol:
```diff
@@ -0,0 +1,32 @@
+    // "profit from corruption", or the maximum amount of margin currency that a token sponsor
+    // could extract from the contract through corrupting the
+    // price feed in their favor.
+    function computeRegularFee(uint startTime, uint endTime, FixedPoint.Unsigned calldata pfc, bytes32 identifier) 
```
Same as above.

mrice32 commented on v1/test/Store.js:
```diff
@@ -0,0 +1,63 @@
+  });
+
+  it("Compute fees", async function() {
+    //Set fee to 10%
```
nit: prefer space between `//` and start of comment here and elsewhere.

mrice32 commented on v1/contracts/Store.sol:
```diff
@@ -0,0 +1,68 @@
+    }
+
+    // Sets a new Oracle fee per second.
+    //TODO change permissioning
```
IMO, we should add permissioning to this method in this PR. WDYT?

mrice32 commented on v1/contracts/Store.sol:
```diff
@@ -0,0 +1,68 @@
+
+    // Sets a new Oracle fee per second.
+    //TODO change permissioning
+    function setFixedOracleFeePerSecond(uint newOracleFee) external {
```
I think this should take in a fixed point number.

mrice32 commented on v1/contracts/Store.sol:
```diff
@@ -0,0 +1,68 @@
+    uint private constant FP_SCALING_FACTOR = 10**18;
+
+    uint private constant WEEKLY_DELAY_FEE = 0; //<-- governance vote?
+    private mapping(address => FixedPoint.Unsigned) finalFees;
```
Is there any way to set the final fee for a currency?

mrice32 commented on v1/contracts/Store.sol:
```diff
@@ -0,0 +1,68 @@
+    uint private fixedOracleFeePerSecond; // Percentage of 10^18. E.g., 1e18 is 100% Oracle fee.
+    uint private constant FP_SCALING_FACTOR = 10**18;
+
+    uint private constant WEEKLY_DELAY_FEE = 0; //<-- governance vote?
```
I think we probably want this to be similar to `fixedOracleFeePerSecond` (set by some specific address managed by `MultiRole`). If we were to add governance, we would just give a governance contract permission to change this.

mrice32 commented on v1/contracts/Store.sol:
```diff
@@ -0,0 +1,68 @@
+    using FixedPoint for uint;
+
+    uint private fixedOracleFeePerSecond; // Percentage of 10^18. E.g., 1e18 is 100% Oracle fee.
+    uint private constant FP_SCALING_FACTOR = 10**18;
```
Ideally, if we use the `FixedPoint` library for all of our fixed point numbers, we shouldn't need this constant anymore.

rosalindstengle replied:
Hmm I think there is still some scaling factor off somewhere because I ended up with test cases with way too many zeros. I will dig into it

rawfalafel commented on v1/contracts/Withdrawable.sol:
```diff
@@ -0,0 +1,22 @@
+import "openzeppelin-solidity/contracts/ownership/Ownable.sol";
+
+
+contract Withdrawable is Ownable {
```
This should be in `common/contracts`, no?

Although if @ptare already started working on moving `v0` to its own branch, doesn't really matter.

rosalindstengle commented:
Added in a governance role that we can sort later, moved many of the values over to fp and updated the FixedPoint library to allow comparison to uint.

mrice32 commented on core/migrations/6_deploy_store.js:
```diff
@@ -0,0 +1,12 @@
+  const store = await deployAndGet(deployer, Store, { from: keys.store });
+  await addToTdr(store, network);
+
+  //TODO set oracle fees & constants
```
nit: add space

mrice32 commented on core/contracts/Store.sol:
```diff
@@ -0,0 +1,112 @@
+        Governance
+    }
+
+    uint private fixedOracleFeePerSecond; // Percentage of 10^18. E.g., 1e18 is 100% Oracle fee.
```
Shouldn't this be a fixed point number since it's a percentage represented as a decimal (i.e. 0.5 => 50%)?

rosalindstengle replied:
Yeah that does make sense. I ran into some trouble making a global variable a part of a struct, but I can mess around with that more.

mrice32 commented on core/test/Store.js:
```diff
@@ -0,0 +1,66 @@
+    store = await Store.new();
+  });
+
+  it("Compute fees", async function() {
```
I lot of the checks below are pretty isolated from one another. WDYT about separating as many as possible into distinct test cases? We've been bad about this in the past, but it would be nice to make our tests more readable :). Even one line test cases could totally make sense!

For example, this could be its own test case:
```
    // Change fee to 20%
    newFee = { value: web3.utils.toWei("0.2", "ether") };
    await store.setFixedOracleFeePerSecond(newFee, { from: owner });

    // Run time tests again
    fees = await store.computeRegularFee(100, 101, pfc, identifier);
    assert.equal(fees.regularFee.toString(), web3.utils.toWei("0.4", "ether"));

    fees = await store.computeRegularFee(100, 110, pfc, identifier);
    assert.equal(fees.regularFee.toString(), web3.utils.toWei("4", "ether"));
```

This could also be its own test case:
```
    // Disallow endTime < startTime.
    assert(await didContractThrow(store.computeRegularFee(2, 1, 10, identifier)));
```

mrice32 commented on core/test/Store.js:
```diff
@@ -0,0 +1,66 @@
+  //Add test final fee for test identifier
+
+  beforeEach(async function() {
+    store = await Store.new();
```
Since you're migrating the store, why not use `store.deployed()`?

mrice32 replied:
Bump

mrice32 commented on core/test/Store.js:
```diff
@@ -0,0 +1,66 @@
+
+  const identifier = web3.utils.utf8ToHex("id");
+
+  //Add test final fee for test identifier
```
Misplaced comment or a TODO?

mrice32 commented on core/test/Store.js:
```diff
@@ -0,0 +1,66 @@
+  it("Compute fees", async function() {
+    // Set fee to 10%
+    let newFee = { value: web3.utils.toWei("0.1", "ether") };
+    let txReciept = await store.setFixedOracleFeePerSecond(newFee, { from: owner });
```
Is this receipt being used anywhere? Usually you'd only need to use a receipt in a test to check that a particular event fired.

mrice32 commented on core/test/Store.js:
```diff
@@ -0,0 +1,66 @@
+    let highFee = { value: web3.utils.toWei("1", "ether") };
+    assert(await didContractThrow(store.setFixedOracleFeePerSecond(highFee, { from: owner })));
+
+    // TODO Check that only permitted role can change the fee
```
Should we add role tests in this PR or do you want to do that in a follow-up?

rosalindstengle replied:
There's a number of more types of testing I plan to add, so I'll do them all in a new PR

mrice32 commented on core/contracts/Store.sol:
```diff
@@ -0,0 +1,112 @@
+    }
+    
+    /*
+     * @notice Do not call this function externally.
```
Can you add these types of comments on all the other methods. If the method is an override of the interface, you only need a docstring at the interface level, and it can be left out for the override.

mrice32 replied:
Bump

rosalindstengle replied:
wait like comments with the @ ? I am a little confused what's missing

ptare commented on core/contracts/Store.sol:
```diff
@@ -0,0 +1,112 @@
+
+    function setFinalFee(address currency, uint finalFee) 
+        external 
+        onlyRoleHolder(uint(Roles.Governance))
```
drive-by:
I think there's some chance that this method may be called much more frequently than `setFixedOracleFeePerSecond`, because IIRC, the final fee adjusts for the relative prices of margin currencies, which could be very volatile.

Thoughts on allowing a role separate from the governance role to set fixed fees?

rosalindstengle replied:
You're probably right about this being accessed more. Wouldn't it still be a governance vote though?

mrice32 replied:
Good idea!

mrice32 replied:
It might be by governance vote at some point in the future. But in the meantime, we would be changing this with a private key. We would probably want a way to delegate this to a "hot" key. I think this same setup might make sense for the setter of the WeeklyDelayFee and the FixedOracleFee as well. We want our "super root" key to be used as little as possible.

mrice32 commented:
It looks like everything was added twice?

mrice32 commented on core/test/Store.js:
```diff
@@ -0,0 +1,70 @@
+    store = await Store.new();
+  });
+
+  it("Compute fees basic check", async function() {
```
Thanks for splitting up these tests!

mrice32 commented on core/contracts/Store.sol:
```diff
@@ -0,0 +1,119 @@
+        Governance
+    }
+
+    FixedPoint.Unsigned private fixedOracleFeePerSecond; // Percentage of 10^18. E.g., 1e18 is 100% Oracle fee.
```
I think the comment needs to be reworded - "1 is a 100% Oracle fee".

mrice32 commented on core/contracts/FixedPoint.sol:
```diff
@@ -28,11 +28,31 @@ library FixedPoint {
         return a.value > b.value;
     }
 
+    /** @dev Whether `a` is less than `b`. */
```
nit: this and the next comment both need to be updated.

mrice32 commented on core/contracts/Store.sol:
```diff
@@ -0,0 +1,119 @@
+import "./MultiRole.sol";
+
+
+// An implementation of StoreInterface that can accept Oracle fees in ETH or any arbitrary ERC20 token.
```
Please update to the same comment style as the methods (might make sense to add an `@title`).

mrice32 commented on core/test/Withdrawable.js:
```diff
@@ -1,89 +0,0 @@
-const { didContractThrow } = require("../../common/SolidityTestUtils.js");
```
Why did you delete this test?

mrice32 commented on core/contracts/Store.sol:
```diff
@@ -0,0 +1,121 @@
+     * @dev Sets a new oracle fee per second
+     */ 
+    function setFixedOracleFeePerSecond(FixedPoint.Unsigned memory newOracleFee) 
+        public 
```
nit: why not `external`?

rosalindstengle replied:
making this function external would mean the param would need to be calldata not memory which upsets the compiler

mrice32 replied:
Oh right - good point.

mrice32 commented on core/contracts/Store.sol:
```diff
@@ -0,0 +1,121 @@
+     * @dev Sets a new final fee for a particular currency
+     */ 
+    function setFinalFee(address currency, FixedPoint.Unsigned memory finalFee) 
+        public 
```
nit: Why not `external`?

mrice32 commented on core/contracts/FixedPoint.sol:
```diff
@@ -28,11 +28,31 @@ library FixedPoint {
     }
 
+    /** @dev Whether `a` is greater than `b`. */
+    function isGreaterThan(Unsigned memory a, uint b) internal pure returns (bool) {
```
If you don't want to add tests for these in this PR, can we add a TODO to add them in a follow up?
